symforce/slam/imu_preintegration/manifold_symbolic_pim.py

```diff
@@ -0,0 +1,309 @@
+
+import symforce.symbolic as sf
+from symforce import typing as T
+from symforce.jacobian_helpers import tangent_jacobians
+
+def internal_imu_residual(
+    rot_i: sf.Rot3,
+    vel_i: sf.V3,
+    pos_i: sf.V3,
+    rot_j: sf.Rot3,
+    vel_j: sf.V3,
+    pos_j: sf.V3,
+    gyro_bias_j: sf.V3,
+    accel_bias_j: sf.V3,
+    # Preintegrated measurements: state
+    DR: sf.Rot3,
+    Dv: sf.V3,
+    Dp: sf.V3,
+    # Other pre-integrated quantities
+    sqrt_info: sf.M99,
+    # put this in 9x6? Would be cleaner
+    DR_D_gyro_bias: sf.M33,
+    Dp_D_accel_bias: sf.M33,
+    Dp_D_gyro_bias: sf.M33,
+    Dv_D_accel_bias: sf.M33,
+    Dv_D_gyro_bias: sf.M33,
+    # other
+    accel_bias_hat: sf.V3,
+    gyro_bias_hat: sf.V3,
+    gravity: sf.V3,
+    dt: T.Scalar,
+    epsilon: T.Scalar,
+) -> sf.Matrix:
+    """
+    An internal helper function to calculate a sort of between factor from the orientation,
+    velocity, and position at one time step and at another time step given, where the expected
+    difference is calculated from the preintegrated IMU measurements between those time steps
+
+    NOTE: If you are looking for a residual for an IMU factor, do not use this. Instead use
+    the one found in symforce/slam/imu_preintegration/imu_factor.h.
+
+    Args:
+        rot_i (sf.Rot3): Orientation at time step i
+        vel_i (sf.V3): Velocity at time step i
+        pos_i (sf.V3): Position at time step i
+        rot_j (sf.Rot3): Orientation at time step j
+        vel_j (sf.V3): Velocity at time step j
+        pos_j (sf.V3): Position at time step j
+        gyro_bias_j (sf.V3): The bias of the gyroscope measurements between timesteps i and j
+        accel_bias_j (sf.V3): The bias of the accelerometer measurements between timesteps i and j
+        DR (sf.Rot3): Preintegrated estimate for rot_i.inverse() * rot_j
+        Dv (sf.V3): Preintegrated estimate for vel_j - vel_i in the body frame at timestep i
+        Dp (sf.V3): Preintegrated estimate for pos change (before velocity and gravity corrections) 
+            in the body frame at timestep i: R_i^T (p_j - p_i - v_i \Delta t - 1/2 g \Delta t^2)
+        sqrt_info (sf.M99): sqrt info matrix of DR('s tangent space), Dv, Dp
+        DR_D_gyro_bias (sf.M33): Derivative of DR w.r.t. gyro_bias evaluated at gyro_bias_hat
+        Dp_D_accel_bias (sf.M33): Derivative of Dp w.r.t. accel_bias evaluated at accel_bias_hat
+        Dp_D_gyro_bias (sf.M33): Derivative of Dp w.r.t. gyro_bias evaluated at gyro_bias_hat
+        Dv_D_accel_bias (sf.M33): Derivative of Dv w.r.t. accel_bias evaluated at accel_bias_hat
+        Dv_D_gyro_bias (sf.M33): Derivative of Dv w.r.t. gyro_bias evaluated at gyro_bias_hat
+        accel_bias_hat (sf.V3): The bias of the accelerometer measurements used during pre-integration
+        gyro_bias_hat (sf.V3): The bias of the gyroscope measurements used during pre-integration
+        gravity (sf.V3): gravity (in the same frame as rot_x, vel_x, and pos_x)
```

Comment:
Hope this can be clarified a bit more. For example, saying that gravity in the inertial frame is [0, 0, -G]. Because some papers use [0, 0, G] instead.